src/main/java/org/testng/internal/reflect/ReflectionRecipes.java

```diff
@@ -331,51 +331,60 @@ public static boolean lenientMatch(final Class<?>[] classes, final Object[] args
-    if (filters != null && !filters.isEmpty()) {
-      final ArrayList<Object> arguments = new ArrayList<>(args.length);
-      final ListBackedImmutableQueue<Object> queue = new ListBackedImmutableQueue<>(args);
-      boolean firstMethodInjected = false;
-      for (final Parameter parameter : parameters) {
-        boolean inject = false;
-        Object injectObject = null;
-        for (final InjectableParameter injectableParameter : filters) {
-          inject = canInject(parameter, injectableParameter);
-          switch (injectableParameter) {
-            case CURRENT_TEST_METHOD:
-              injectObject = injectionMethod;
-              if (inject && !firstMethodInjected) {
-                firstMethodInjected = true;
-              } else {
-                inject = false;
-              }
-              break;
-            case ITEST_CONTEXT:
-              injectObject = context;
-              break;
-            case ITEST_RESULT:
-              injectObject = testResult;
-              break;
-            case XML_TEST:
-              injectObject = context != null ? context.getCurrentXmlTest() : null;
-              break;
-            default:
-              break;
-          }
-          if (inject) {
-            arguments.add(injectObject);
+    if (filters == null || filters.isEmpty()) {
+      return args;
+    }
+    final ArrayList<Object> arguments = new ArrayList<>(args.length);
+    final ListBackedImmutableQueue<Object> queue = new ListBackedImmutableQueue<>(args);
+    boolean firstMethodInjected = false;
+    for (final Parameter parameter : parameters) {
+      boolean inject = false;
+      Object injectObject = null;
+      for (final InjectableParameter injectableParameter : filters) {
+        inject = canInject(parameter, injectableParameter);
+        switch (injectableParameter) {
+          case CURRENT_TEST_METHOD:
+            injectObject = injectionMethod;
+            if (inject && !firstMethodInjected) {
+              firstMethodInjected = true;
+            } else {
+              inject = false;
+            }
+            break;
+          case ITEST_CONTEXT:
+            injectObject = context;
+            break;
+          case ITEST_RESULT:
+            injectObject = testResult;
+            break;
+          case XML_TEST:
+            injectObject = context != null ? context.getCurrentXmlTest() : null;
+            break;
+          default:
             break;
-          }
         }
-
-        if (!inject) {
-          arguments.add(queue.poll());
+        if (inject) {
+          arguments.add(injectObject);
+          break;
         }
       }
-      final Object[] injectedArray = new Object[arguments.size()];
-      return arguments.toArray(injectedArray);
-    } else {
-      return args;
+
+      if (!inject) {
+        arguments.add(queue.poll());
+      }
     }
+    if (!queue.backingList.isEmpty()) { // TODO Remove duplication from Paramters#getInjectedParameter
```

Comment:
@juherr - No I haven't yet thought about it. I thought it would be better if we have it as a separate PR rather than club those changes in here.